### DISCUSSION TITLE
openEO EODC backend deployment via TOSCA

### DISCUSSION THREAD

**Discussion by andrea-manzi** [interTwin-eu/repository-template#6] [General]
Cluster deployed via IM at https://openeo.62.217.83.68.nip.io/1.1.0/ 

Need to clarify the EGI Check-in used. I have setup via IM to use the EGI Check-in demo instance, is that automatically configured ? or only the Production instance is there by default?

**Comment by clausmichele** [comment]
I report here the exampels I shared via email to test the deployed back-end:

> Anyway, the Eurac openEO development back-end uses the same code base (openeo-processes-dask) for the processes implementation, and this works against it (you can also try it):

```python
import openeo
from openeo.metadata import CollectionMetadata
from openeo.processes import array_element

conn = openeo.connect('https://dev.openeo.eurac.edu').authenticate_basic("guest","changeme")

url = https://earth-search.aws.element84.com/v1/collections/sentinel-2-l2a
spatial_extent =  {"east": 11.40, "north": 46.52, "south": 46.46, "west": 11.25}
temporal_extent = ["2022-06-01", "2022-06-30"]
bands = ["red", "nir"]
s2_datacube = conn.load_stac(
    url=url,
    spatial_extent=spatial_extent,
    temporal_extent=temporal_extent,
    bands=bands
)

s2_datacube.metadata = CollectionMetadata({"cube:dimensions":{"time":{"type":"temporal"},"band":{"type":"bands"}}})

def NDVI(data):    
    b04 = data.array_element(label="red")
    b08 = data.array_element(label="nir")
    return (b08 - b04) / (b08 + b04)

ndvi = s2_datacube.reduce_dimension(dimension="band",reducer=NDVI)
ndvi_median = ndvi.reduce_dimension(dimension="time", reducer="median")
result_ndvi = ndvi_median.download("ndvi_eurac.nc")
```

> When trying to use the deployed InterTwin openEO back-end I get the following error. It seems like it’s requiring authentication, which I’m unaware of.
> 
> Code to try:

```python
import openeo
import requests
from openeo.metadata import CollectionMetadata
from openeo.processes import array_element

s = requests.Session()
s.verify = False
conn = openeo.connect('https://openeo.62.217.83.68.nip.io/1.1.0/',session=s)

url = https://earth-search.aws.element84.com/v1/collections/sentinel-2-l2a
spatial_extent =  {"east": 11.40, "north": 46.52, "south": 46.46, "west": 11.25}
temporal_extent = ["2022-06-01", "2022-06-30"]
bands = ["red", "nir"]
s2_datacube = conn.load_stac(
    url=url,
    spatial_extent=spatial_extent,
    temporal_extent=temporal_extent,
    bands=bands
)

s2_datacube.metadata = CollectionMetadata({"cube:dimensions":{"time":{"type":"temporal"},"band":{"type":"bands"}}})

def NDVI(data):    
    b04 = data.array_element(label="red")
    b08 = data.array_element(label="nir")
    return (b08 - b04) / (b08 + b04)

ndvi = s2_datacube.reduce_dimension(dimension="band",reducer=NDVI)
ndvi_median = ndvi.reduce_dimension(dimension="time", reducer="median")
result_ndvi = ndvi_median.download("ndvi_intertwin.nc")
```

> error:
```
OpenEoApiPlainError: [422] {"detail":[{"loc":["header","authorization"],"msg":"field required","type":"value_error.missing"}]}
```

It would be nice to have a working example using batch jobs and getting the result in STAC. @SerRichard where will the result be stored? Is there a linked object storage to this openEO back-end?

@andrea-manzi I don't see Sean in the members, could you add him so that he can take part to the discussion?

**Reply by andrea-manzi** [reply]
I have invited Sean to the organization !

**Reply by SerRichard** [reply]
@clausmichele Results currently would be stored in the volume claim used by the backend deployment, once we have these examples working it might be worth discussing making them available via S3 instead.

At the minute I think the auth issue in the example above is related to the credentials/oidc endpoint not being available, this was an oversight on my side as I tested the urls VIA postman and not the client, so I just sent process graphs to the endpoint with the header provided.

I will update the codebase and  check this notebook will work on a deployment on myside before providing the next image update to @micafer

**Reply by SerRichard** [reply]
The initial issue here was the authorization. Following on I had some further issues with the result saving on the xarrays from stackstac and then a mistake I had made in the result return.

I hadn't anticipated, but when you load items with stackstac, it seems to add all of the item properties to the xarray object, some properties go into the attributes ( in the form of a RasterSpec object which they define) , but then a lot of properties get dumped into the xarrays coordinates ( and not added to the dimensions? ). When trying to write these xarrays to netcdf, the majority of these additional values were unserializable. As a result I've dropped all coordinate values that are not also used in the dimensions of the xarray, and dropped the RasterSpec from the attributes.

Additionally, I had made the mistake of only providing tar results via sync jobs, I've updated this to reflect the API documentation, that single files are returned directly, and only multiple files are returned as tar.

```
# Single file
result_ndvi = ndvi_median.download("ndvi_intertwin.nc")

# Multi file
result_ndvi = ndvi_median.download("ndvi_intertwin.tar", format="netcdf")
```

The API has been updated to handle these different cases. Though currently STAC output is not included as part of the tar result, which is an openeo-api spec recommendation, @andrea-manzi  we can add this quite easily if we want.

**Reply by clausmichele** [reply]
@SerRichard yea I am aware of the properties handled by stackstac, which are not so compatible with netCDF. For that and other reasons we are currently using/testing odc-stac, as described here: https://github.com/Open-EO/openeo-processes-dask/issues/235
I am also working on improving the support of Zarr and Kerchunk and soon include them in a PR.

**Comment by micafer** [comment]
Hi @clausmichele 

If you change the line:

```py
conn = openeo.connect('https://openeo.62.217.83.68.nip.io/1.1.0/',session=s)
```

With this, the authentication works (but I get another errors)
```py
token = "ey.........."
conn = openeo.connect('https://openeo.62.217.83.68.nip.io/1.1.0',session=s).authenticate_oidc_access_token(token)
```

To get the value of the ``token`` variable you can go here, and copy the value of the Access Token:
https://aai-demo.egi.eu/token/

**Reply by clausmichele** [reply]
For me, this returns the following error:

```python
import openeo
import requests

token = "ey..."
s = requests.Session()
s.verify = False
conn = openeo.connect('https://openeo.62.217.83.68.nip.io/1.1.0',session=s).authenticate_oidc_access_token(token)

>> OpenEoApiPlainError: [404] {"detail":"Not Found"}
```

**Reply by SerRichard** [reply]
For some reason the /credentials/oidc url isn't being provided by the server, will check why and release a new image if needed

**Reply by SerRichard** [reply]
I've updated the openeo-fastapi package to include the /credentials/oidc endpoint as a default. Logging in now seems to work in both the following ways.

```
# Traditional device flow ( tested using the production egi deployment )
conn = openeo.connect('http://127.0.0.1:8000/1.1.0/', session=s).authenticate_oidc()
conn

# Using the authenticate_oidc_access_token directly
conn = openeo.connect('http://127.0.0.1:8000/1.1.0/', session=s)
conn.authenticate_oidc_access_token(token, "egi")

# .authenticate_oidc_access_token() Doesn't return anything, so the following will always lead to conn being None.
conn = openeo.connect('http://127.0.0.1:8000/1.1.0/', session=s).authenticate_oidc_access_token(token, "egi")
```

@micafer, I've updated the oidcRoles setting to oidcPolicies , an empty string can be provided for now. The new oidcPolicies settings is a general way of determining what checks need to be carried out against a userinfo response body from the issuer, I.e, checking the user belongs to a keycloak group, or with egi, whether they have a specific eduperson_entitlment.  If / when the policy setting wants to be used, it works with both egi and keycloak and is covered in the documentation.

**Reply by micafer** [reply]
I updated the TOSCA artifacts to use this new name.

**Reply by micafer** [reply]
Hi @SerRichard ,

I have a problem with the deployment, the openeo-openeo-argo pod is failing with this error:

```py
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/openeo_argoworkflows_api/.venv/lib/python3.11/site-packages/uvicorn/_subprocess.py", line 78, in subprocess_started
    target(sockets=sockets)
  File "/opt/openeo_argoworkflows_api/.venv/lib/python3.11/site-packages/uvicorn/server.py", line 65, in run
    return asyncio.run(self.serve(sockets=sockets))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/opt/openeo_argoworkflows_api/.venv/lib/python3.11/site-packages/uvicorn/server.py", line 69, in serve
    await self._serve(sockets)
  File "/opt/openeo_argoworkflows_api/.venv/lib/python3.11/site-packages/uvicorn/server.py", line 76, in _serve
    config.load()
  File "/opt/openeo_argoworkflows_api/.venv/lib/python3.11/site-packages/uvicorn/config.py", line 433, in load
    self.loaded_app = import_from_string(self.app)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/openeo_argoworkflows_api/.venv/lib/python3.11/site-packages/uvicorn/importer.py", line 19, in import_from_string
    module = importlib.import_module(module_str)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/opt/openeo_argoworkflows_api/app.py", line 10, in <module>
    from openeo_argoworkflows_api.jobs import ArgoJobsRegister
  File "/opt/openeo_argoworkflows_api/jobs.py", line 35, in <module>
    from openeo_argoworkflows_api.tasks import queue_to_submit, submit_job
  File "/opt/openeo_argoworkflows_api/tasks.py", line 13, in <module>
    settings = ExtendedAppSettings()
               ^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/env_settings.py", line 40, in pydantic.env_settings.BaseSettings.__init__
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for ExtendedAppSettings
OIDC_POLICIES
  Policy '' contains too many comma seperated values. (type=value_error)
``

I tried several values of the variable and it is failing in all the cases.

**Reply by andrea-manzi** [reply]
@micafer is this blocking for the video for the Review? and in case shall we go back to the previous version?

**Reply by micafer** [reply]
I can record the deployment, but I cannot access the service because it is not running.

**Reply by micafer** [reply]
Tomorrow I can try to record it using the previous version.

**Reply by micafer** [reply]
I created an issue in the OpenEO API to track this problem:
https://github.com/eodcgmbh/openeo-fastapi/issues/59

**Comment by SerRichard** [comment]
@micafer The deployment can be updated, if you'd refer to the latest changes in the chart, the issues above should get resolved. The latest tags for the api and executor have also changed.

Currently there's two points of open work that I'd like to address.

 1. Update helm chart to allow for disabling the installation of chart dependencies and reusing existing deployments.
 2.  Tiling the outputs of save result.

The first is pretty self explanatory, and I just need to allocate time to complete. The second, I'm still open for suggestions, but it would be good to implement some default tiling grid to control the output of the filesize, to limit possible pressure on the api on download, and on the dask workers when writing results.

**Reply by micafer** [reply]
@andrea-manzi  could you give me access to the current one to update it?
or you can try to deploy a new one.

**Reply by andrea-manzi** [reply]
i sent you the key to acccess

**Comment by micafer** [comment]
Hi all,

After the last update of the API it seems that it is working now using this code to authenticate:

```py
import openeo
import requests

s = requests.Session()
token = "SOME_TOKEN"
conn = openeo.connect('https://openeo.intertwin.fedcloud.eu/1.1.0/', session=s)
conn.authenticate_oidc_access_token(token, "egi")

...
```

I tied the computation suggested by @clausmichele, but it gets an error:

```py
Failed to parse API error response: [500] 'Internal Server Error' (headers: {'Date': 'Thu, 24 Oct 2024 14:41:23 GMT', 'Content-Type': 'text/plain; charset=utf-8', 'Content-Length': '21', 'Connection': 'keep-alive', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'})
```

And @SerRichard in the logs of the `openeo-argo-queue-worker` I saw this error:

```py
hera.exceptions.Forbidden: Server returned status code 403 with message: `workflows.argoproj.io is forbidden: User "system:serviceaccount:openeo:openeo-argo-access-sa" cannot create resource "workflows" in API group "argoproj.io" in the namespace "openeo"`
```

**Reply by SerRichard** [reply]
@micafer For login via the openeo client, you would want to do something like 
```# Traditional device flow
conn = openeo.connect('http://127.0.0.1:8000/1.1.0/', session=s).authenticate_oidc()
```

I can check though on my side, what are the oidc settings deployed?

I am just updating the helm chart now for the "openeo-argo-access-sa", strange this was not an issue before? Did anything change cluster side?

**Reply by micafer** [reply]
The OIDC settings are the default ones.

**Reply by SerRichard** [reply]
Okay, then in that case, it should work with the line provided above. You will be redirect to egi and carry out a device flow. I have just deploy the application in k8s and ran Michele's code, if you get further errors, either the api or executor logs would need to be shared to debug.

The helm chart has been updated, the following verbs have been added to the role bound to the forbidden service account: 
"get","create", "list", "update", "patch", let me know if it persists!

**Reply by micafer** [reply]
Now it seems to work, but I have another problem.
It has been created a dask-worker that requests 4 cpus and 8 GB of RAM, as currently the cluster does not have enough resources.
Is there a way to specify the amount of resources requested by the dask workers?

**Reply by SerRichard** [reply]
Ooooo, interesting

So in the api I have the concept of a dask_profile, that gets parsed the executor, and the executor will make those requests.
i.e.
```
dask_profile = {
          "GATEWAY_URL": settings.DASK_GATEWAY_SERVER,
          "OPENEO_EXECUTOR_IMAGE": settings.OPENEO_EXECUTOR_IMAGE,
          "WORKER_CORES": "4",
          "WORKER_MEMORY": "8",
          "WORKER_LIMIT": "6",
          "CLUSTER_IDLE_TIMEOUT": "3600"
}
```

Ofc, as you can see those values are hard coded. An easy approach would be to make those values configurable via the application settings, and set defaults if values are not provided.

That being said, the idea of the dask profile was to evolve with some idea of a user_profile, i.e if you had a free user vs some tiered user. Is that something that we want to be supported in this project?

Regardless, I can make them immediately available as settings, and then we can move the profile idea forward if that is desired.

**Reply by micafer** [reply]
It will be great if you can make it available as settings.

**Reply by SerRichard** [reply]
A new api image has been released ( api-2024.10.2 ) and the helm chart has been updated, please refer to the read me for the new environment variables that can be set in the values.yaml. This should allow for smaller or larger dask cluster configurations. Any issues, give me another message.

**Reply by micafer** [reply]
Is there a minimum recommendable values to set?

**Comment by micafer** [comment]
In the current cluster I relaunched with a bigger WN and now the jobs fit in the size of the node and the example is completed successfully!

**Reply by clausmichele** [reply]
I also confirm that I was able to run successfully a sample graph in synchronous mode!

**Comment by clausmichele** [comment]
@micafer the openEO back-end is not reachable anymore, it returns 404: https://openeo.intertwin.fedcloud.eu/1.1.0/

**Reply by micafer** [reply]
Hi @clausmichele,
I tested the code again, and it is working.
The URL "ttps://openeo.intertwin.fedcloud.eu/1.1.0/" always has returned 404.

**Reply by clausmichele** [reply]
Ok thanks, it works with the openEO Python client but it doesn't with the openEO Web Editor. Any idea why? @SerRichard maybe?
https://editor.openeo.org/?server=https://openeo.intertwin.fedcloud.eu/1.1.0/
Edit: I see that it redirects to: https://openeo.intertwin.fedcloud.eu/openeo/1.1.0/, but this also doesn't work with the Web Editor.

**Reply by clausmichele** [reply]
The issues comes from the missing CORS access-control-allow-origin header.

**Reply by andrea-manzi** [reply]
is this needed on the openEO REST API? or it can be added in the kubernetes ingress?

**Reply by micafer** [reply]
I have added in the Ingress @clausmichele could you check again?

**Reply by clausmichele** [reply]
![image](https://github.com/user-attachments/assets/1e6eb3db-aaff-4297-bf46-54e9016add6e)

Now it can connect but doesn't show the possibility to authenticate

**Reply by andrea-manzi** [reply]
how this is supposed to work? cause indeed if the login should be handled by the openEO backend we need a EGI Check-in client id to be created

**Reply by SerRichard** [reply]
@clausmichele Apologies, I was away for the past week. The web-editor is making an unsuccessful request to "https://openeo.intertwin.fedcloud.eu/openeo/1.1.0/.well-known/openeo" instead of "https://openeo.intertwin.fedcloud.eu/.well-known/openeo". In the EODC deployment if I remember correctly we just made a redirect available in the fastapi to handle this. I think that should resolve the issue. This could be tested with a redirect in the ingress, but I have also noticed a small typo in the "/credentials/oidc" endpoint, so I can update the two of these at the same time.

@andrea-manzi The backend takes tokens and checks them against the configured provider. The editor handles the login flow using the oidc provider specified by the backend. The user gets directed to the provider to login, and back the editor with a token to use against the api afterward.

**Reply by SerRichard** [reply]
@micafer I've just upated the api image, would you update the deployment so we can check if this has resolved the issue? If not, I'll take a closer look!

Tag: api-2024.11.1

**Reply by micafer** [reply]
cluster updated to version 11.1

**Reply by SerRichard** [reply]
Still the same, I'll redeploy this on my side and see if I'm able to resolve it today

**Reply by SerRichard** [reply]
@micafer please update to image api-2024.11.5.

After a few releases and some testing with the editor, I realised the issue was in the openeo-fastapi package. The endpoints on the top level of the application were not listed in the capabilities response, which the editor uses to infer the location of "/credentials/oidc".

**Reply by micafer** [reply]
updated to version 11.5

**Reply by SerRichard** [reply]
Greaat, login is working! Seems to be an issue with the "/files" endpoints, will take a look at the logs on my side and see what the issue is, as that feature should be supported!

**Reply by SerRichard** [reply]
@micafer Was trickier to get enabled that I suspected, with the latest tag "api-2024.11.10", the files endpoints are now all functioning as expected and working with the editor

**Reply by micafer** [reply]
updated!

**Comment by andrea-manzi** [comment]
i guess now everything is working as expected, closing
